### PR TITLE
[SPARK-49531][PYTHON][CONNECT] Support line plot with plotly backend

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Support line plot with plotly backend on both Spark Connect and Spark classic.

### Why are the changes needed?
While Pandas on Spark supports plotting, PySpark currently lacks this feature. The proposed API will enable users to generate visualizations, such as line plots, by leveraging libraries like Plotly. This will provide users with an intuitive, interactive way to explore and understand large datasets directly from PySpark DataFrames, streamlining the data analysis workflow in distributed environments.

See more at [PySpark Plotting API Specification](https://docs.google.com/document/d/1IjOEzC8zcetG86WDvqkereQPj_NGLNW7Bdu910g30Dg/edit?usp=sharing) in progress.

Part of https://issues.apache.org/jira/browse/SPARK-49530.

### Does this PR introduce _any_ user-facing change?
Yes.

```python
>>> data = [("A", 10, 1.5), ("B", 30, 2.5), ("C", 20, 3.5)]
>>> columns = ["category", "int_val", "float_val"]
>>> sdf = spark.createDataFrame(data, columns)
>>> sdf.show()
+--------+-------+---------+                                                    
|category|int_val|float_val|
+--------+-------+---------+
|       A|     10|      1.5|
|       B|     30|      2.5|
|       C|     20|      3.5|
+--------+-------+---------+

>>> f = sdf.plot(kind="line", x="category", y="int_val")
>>> f.show()  # see below
>>> g = sdf.plot.line(x="category", y=["int_val", "float_val"])
>>> g.show()  # see below
```
`f.show()`:
newplot

`g.show()`:
newplot (1)

### How was this patch tested?
Unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by xinrong-meng** [comment]
@zhengruifeng @HyukjinKwon may I get a review please?

**Comment by xinrong-meng** [comment]
@zhengruifeng certainly, added to the pr description, thanks!

**Comment by xinrong-meng** [comment]
Failed test is irrelevant, retriggered:
```
FAIL [1.024s]: test_not_found_fails (pyspark.sql.tests.connect.client.test_client.SparkConnectClientReattachTestCase.test_not_found_fails)
SPARK-48056: Assert that the client fails from session or operation not found error
```

**Review by zhengruifeng — COMMENTED** [review]
In general LGTM

File: python/pyspark/sql/tests/connect/test_parity_frame_plot.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+from pyspark.testing.connectutils import ReusedConnectTestCase
+from pyspark.sql.tests.plot.test_frame_plot import DataFramePlotTestsMixin
+
+
+class FramePlotParityTests(DataFramePlotTestsMixin, ReusedConnectTestCase):
+    pass
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (xinrong-meng) [reply python/pyspark/sql/tests/connect/test_parity_frame_plot.py:26]: Good catch! Added.

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged to master

**Comment by xinrong-meng** [comment]
Thank you @zhengruifeng !

**Review by dongjoon-hyun — COMMENTED** [review]
Unfortunately, this seems to break a few daily CIs, @xinrong-meng .
- https://github.com/apache/spark/actions/workflows/build_python_pypy3.9.yml

```
======================================================================
ERROR: test_backend (pyspark.sql.tests.plot.test_frame_plot.DataFramePlotTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/sql/utils.py", line 124, in require_minimum_plotly_version
    import plotly  # noqa: F401
ModuleNotFoundError: No module named 'plotly'

The above exception was the direct cause of the following exception:
```

- https://github.com/apache/spark/actions/workflows/build_python_connect.yml

```
ModuleNotFoundError: No module named 'pyspark.sql.plot'

Had test failures in pyspark.ml.tests.connect.test_connect_classification with python3; see logs.
```

File: python/pyspark/sql/plot/core.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Any, TYPE_CHECKING, Optional, Union
+from types import ModuleType
+from pyspark.errors import PySparkRuntimeError, PySparkValueError
+from pyspark.sql.utils import require_minimum_plotly_version
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/plot/core.py:21]: This PR looks like `plotly` as mandatory at `pyspark/sql/plot` module.

**Comment by dongjoon-hyun** [comment]
To @xinrong-meng and @zhengruifeng , 
- For the import issues, we can elevate `plotly` to the mandatory SQL if you want. However, the document and `requirement.txt` and CIs seems to be out of order currently.
- For Spark Connect CI's `ModuleNotFoundError: No module named 'pyspark.sql.plot'`, I have no clue. Since this PR aims to support both `Classic` and `Connect`, I guess the failure looks legitimate to me.

Let me revert this for now. I hope we can bring this back (with revision) after `Preview2`.

**Comment by dongjoon-hyun** [comment]
This is reverted via
- https://github.com/apache/spark/commit/fa6a0786bb4b23a895e68a721df9ee88684c4fab

**Comment by dongjoon-hyun** [comment]
For the record, both CIs are back to green. I'm going to cut `4.0.0-preview2 RC1` soon.
- https://github.com/apache/spark/actions/workflows/build_python_pypy3.9.yml
- https://github.com/apache/spark/actions/workflows/build_python_connect.yml

**Comment by xinrong-meng** [comment]
Sorry for the inconvenience caused, and thanks for handling that @dongjoon-hyun! I’ll follow up with the following after 4.0.0-preview2:

- Elevate Plotly to the mandatory SQL and the reverted commit.
- Ensure the CI passes.

May I ask if build_python_pypy and build_python_connect are part of the PR builders? If not, how do I trigger them to verify?

**Comment by dongjoon-hyun** [comment]
Thank you, @xinrong-meng . They are daily CIs, not a part of PR builders.
It should be verified manually. Before testing them, please test in your environment after uninstalling `plotly`. Then, it will be similar to you. (at last for `build_python_pypy3.9`).

File: python/pyspark/sql/plot/core.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/sql/plot/core.py:21]: we should make `plotly` an optional dependency

**Comment by xinrong-meng** [comment]
Filed https://github.com/apache/spark/pull/48139 as a follow-up.

**Pull request closed without merging**